hoodie-common/src/main/java/com/uber/hoodie/common/table/log/avro/AvroLogAppender.java

```diff
@@ -0,0 +1,135 @@
+import org.apache.avro.generic.IndexedRecord;
+import org.apache.hadoop.fs.AvroFSInput;
+import org.apache.hadoop.fs.FSDataOutputStream;
+import org.apache.hadoop.fs.FileContext;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
+import org.apache.hadoop.hdfs.DistributedFileSystem;
+import org.apache.hadoop.hdfs.protocol.AlreadyBeingCreatedException;
+import org.apache.hadoop.ipc.RemoteException;
+import org.apache.log4j.LogManager;
+import org.apache.log4j.Logger;
+
+import java.io.IOException;
+import java.util.List;
+
+/**
+ * AvroLogAppender appends a bunch of IndexedRecord to a Avro data file.
+ * If auto-flush is set, every call to append writes out a block.
+ * A avro block corresponds to records appended in a single commit.
+ *
+ * @see org.apache.avro.file.DataFileReader
+ */
+public class AvroLogAppender {
+    private final static Logger log = LogManager.getLogger(AvroLogAppender.class);
+    private final HoodieLogAppendConfig config;
+    private FSDataOutputStream output;
+    private DataFileWriter<IndexedRecord> writer;
+    private boolean autoFlush;
+
+    public AvroLogAppender(HoodieLogAppendConfig config) throws IOException, InterruptedException {
+        FileSystem fs = config.getFs();
+        this.config = config;
+        this.autoFlush = config.isAutoFlush();
+        GenericDatumWriter<IndexedRecord> datumWriter =
+            new GenericDatumWriter<>(config.getSchema());
+        this.writer = new DataFileWriter<>(datumWriter);
+        Path path = config.getLogFile().getPath();
+
+        if (fs.exists(path)) {
+            //TODO - check for log corruption and roll over if needed
+            log.info(config.getLogFile() + " exists. Appending to existing file");
+            // this log path exists, we will append to it
+            fs = FileSystem.get(fs.getConf());
+            try {
+                this.output = fs.append(path, config.getBufferSize());
+            } catch (RemoteException e) {
+                // this happens when either another task executor writing to this file died or data node is going down
+                if(e.getClassName().equals(AlreadyBeingCreatedException.class.getName()) && fs instanceof DistributedFileSystem) {
+                    log.warn("Trying to recover log on path " + path);
+                    if(FSUtils.recoverDFSFileLease((DistributedFileSystem) fs, path)) {
+                        log.warn("Recovered lease on path " + path);
+                        // try again
+                        this.output = fs.append(path, config.getBufferSize());
+                    } else {
+                        log.warn("Failed to recover lease on path " + path);
+                        throw new HoodieException(e);
+                    }
+                }
+            }
+            this.writer.appendTo(new AvroFSInput(FileContext.getFileContext(fs.getConf()), path), output);
+            this.writer.setFlushOnEveryBlock(true);
```

Comment:
setFlushOnEveryBlock is slightly different. autoFlush config says should we write out a block after every call to append(List records). setFlushOnEveryBlock - means whenever a block is written out flush to disk - we always want that.